File: 209-minimum-size-subarray-sum/209-minimum-size-subarray-sum.py
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        def isSumEqualGreater(l: int) -> bool:
            total = 0
            for i in range(l):
                total += nums[i]
            
            for j in range(i + 1, len(nums)):
                if total >= target:
                    return True
                
                total -= nums[j - l]
                total += nums[j]
            return total >= target
        
        lo, hi = 0, len(nums)
        while lo < hi:
            mid = lo + (hi - lo) // 2
            if isSumEqualGreater(mid + 1):
                hi = mid
            else:
                lo = mid + 1
        return 0 if lo == len(nums) else lo + 1

File: 209-minimum-size-subarray-sum/209-minimum-size-subarray-sum.java
class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        int lo = 0, hi = nums.length;
        while(lo < hi){
            int mid = lo + (hi - lo) / 2;
            if(isSumEqualGreater(nums, target, mid + 1))
                hi = mid;
            else
                lo = mid + 1;
        }
        return lo == nums.length ? 0 : lo + 1;
    }
    
    private boolean isSumEqualGreater(int[] nums, int target, int l){
        int total = 0;
        for(int i = 0; i < l; i++)
            total += nums[i];
        
        for(int j = l; j < nums.length; j++){
            if (total >= target)
                return true;
            total -= nums[j - l];
            total += nums[j];
        }
        return total >= target;
    }
}